### PR TITLE
[19378] Add XML support for ThreadSettings

### PR DESCRIPTION
**Pull request by EduPonz**
## Description

This PR provides support for configuring Fast DDS ThreadSettings using XML.

## Contributor Checklist
- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- [x] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally 
- [x] Any new/modified methods have been properly documented using Doxygen. 
- [x] Changes are ABI compatible. 
- [x] Changes are API compatible. 
- *N/A*: New feature has been added to the `versions.md` file (if applicable).
- *N/A*: New feature has been documented/Current behavior is correctly described in the documentation. 
    
- *N/A*: Applicable backports have been included in the description.

## Reviewer Checklist
- [x] The PR has a milestone assigned.
- [x] Check contributor checklist is correct.
- [ ] Check CI results: changes do not issue any warning.
- [ ] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

**Comment by EduPonz** [comment]
@richiprosima please test mac

File: resources/xsd/fastRTPS_profiles.xsd [hunk 1]
```diff
@@ -1372,11 +1372,12 @@
...
     <!--QoS Data Sharing:
-        ├ kind         [string] ("AUTOMATIC", "ON", "OFF"),
-        ├ shared_dir   [string],
-        ├ domain_ids   [0~*],
-        |   └ domainID [uint32] (0~232)
-        └ max_domains  [uint32] -->
+        ├ kind                         [string] ("AUTOMATIC", "ON", "OFF"),
+        ├ shared_dir                   [string],
+        ├ domain_ids                   [0~*],
+        |   └ domainID                 [uint32] (0~232)
```
Comment (MiguelCompany) [inline resources/xsd/fastRTPS_profiles.xsd:1380]: Totally off-topic to the PR: Since this is not related with the DDS domain ID, the values here have no limits. They can be considered similar to the partitions.

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

File: src/cpp/rtps/xmlparser/XMLParser.cpp [hunk 2]
```diff
@@ -1961,6 +2175,47 @@ XMLP_ret XMLParser::fillDataNode(
...
+        }
+#if HAVE_SECURITY
+        else if (strcmp(name, SECURITY_LOG_THREAD) == 0)
+        {
+            if (XMLP_ret::XML_OK != getXMLThreadSettings(*p_aux0, participant_node.get()->rtps.security_log_thread))
+            {
+                return XMLP_ret::XML_ERROR;
+            }
+        }
+#endif // if HAVE_SECURITY
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply src/cpp/rtps/xmlparser/XMLParser.cpp:2218]: good catch!

File: resources/xsd/fastRTPS_profiles.xsd (hunk 1 above)
Comment (EduPonz) [reply resources/xsd/fastRTPS_profiles.xsd:1380]: You're absolutely right

File: src/cpp/rtps/participant/RTPSParticipantImpl.cpp [hunk 3]
```diff
@@ -161,6 +161,7 @@ RTPSParticipantImpl::RTPSParticipantImpl(
         UDPv4TransportDescriptor descriptor;
         descriptor.sendBufferSize = m_att.sendSocketBufferSize;
         descriptor.receiveBufferSize = m_att.listenSocketBufferSize;
+        descriptor.default_reception_threads(m_att.builtin_transports_reception_threads);
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply src/cpp/rtps/participant/RTPSParticipantImpl.cpp:164]: 59e3e626a

File: test/unittest/xmlparser/XMLElementParserTests.cpp [hunk 4]
```diff
@@ -4112,3 +4124,206 @@ TEST_F(XMLParserTests, env_var_substitution)
...
+    /* Define the test cases */
+    std::vector<std::pair<std::vector<std::string>, XMLP_ret>> test_cases =
+    {
+        {{"12", "12", "12", "12", ""}, XMLP_ret::XML_OK},
+        {{"-1", "12", "12", "12", ""}, XMLP_ret::XML_OK},
+        {{"12", "-1", "12", "12", ""}, XMLP_ret::XML_OK},
+        {{"12", "12", "12", "-1", ""}, XMLP_ret::XML_OK},
+        {{"-2", "12", "12", "12", ""}, XMLP_ret::XML_ERROR},
+        {{"12", "-2", "12", "12", ""}, XMLP_ret::XML_ERROR},
+        {{"12", "-2", "-1", "12", ""}, XMLP_ret::XML_ERROR},
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply test/unittest/xmlparser/XMLElementParserTests.cpp]: 59e3e626a

File: test/unittest/xmlparser/XMLElementParserTests.cpp [hunk 5]
```diff
@@ -4112,3 +4123,66 @@ TEST_F(XMLParserTests, env_var_substitution)
...
+{
+    /* Define the test cases */
+    std::vector<std::pair<std::vector<std::string>, XMLP_ret>> test_cases =
+    {
+        {{"12", "12", "12", "12", ""}, XMLP_ret::XML_OK},
+        {{"-1", "12", "12", "12", ""}, XMLP_ret::XML_OK},
+        {{"12", "-1", "12", "12", ""}, XMLP_ret::XML_OK},
+        {{"12", "12", "12", "-1", ""}, XMLP_ret::XML_OK},
+        {{"-2", "12", "12", "12", ""}, XMLP_ret::XML_ERROR},
+        {{"12", "-2", "12", "12", ""}, XMLP_ret::XML_ERROR},
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply test/unittest/xmlparser/XMLElementParserTests.cpp]: 59e3e626a

File: test/unittest/xmlparser/XMLElementParserTests.cpp [hunk 6]
```diff
@@ -4112,3 +4123,66 @@ TEST_F(XMLParserTests, env_var_substitution)
...
+        {{"12", "-2", "-1", "12", ""}, XMLP_ret::XML_ERROR},
+        {{"12", "12", "12", "-2", ""}, XMLP_ret::XML_ERROR},
+        {{"a", "12", "12", "12", ""}, XMLP_ret::XML_ERROR},
+        {{"12", "a", "12", "12", ""}, XMLP_ret::XML_ERROR},
+        {{"12", "12", "a", "12", ""}, XMLP_ret::XML_ERROR},
+        {{"12", "12", "12", "a", ""}, XMLP_ret::XML_ERROR},
+        {{"12", "12", "12", "12", "<stack_size>12</stack_size>"}, XMLP_ret::XML_ERROR},
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply test/unittest/xmlparser/XMLElementParserTests.cpp:4149]: 59e3e626a

File: test/unittest/xmlparser/XMLElementParserTests.cpp [hunk 7]
```diff
@@ -4112,3 +4124,206 @@ TEST_F(XMLParserTests, env_var_substitution)
...
+        // Extract ThreadSetting
+        titleElement = xml_doc.RootElement();
+        ASSERT_EQ(expectation, XMLParserTest::getXMLThreadSettings_wrapper(titleElement, thread_settings));
+
+        // Validate in the OK cases
+        if (expectation == XMLP_ret::XML_OK)
+        {
+            ASSERT_EQ(thread_settings.scheduling_policy, static_cast<int32_t>(std::stoi(params[0])));
+            ASSERT_EQ(thread_settings.priority, static_cast<int32_t>(std::stoi(params[1])));
+            ASSERT_EQ(thread_settings.affinity, static_cast<uint32_t>(std::stoi(params[2])));
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply test/unittest/xmlparser/XMLElementParserTests.cpp]: 59e3e626a

File: src/cpp/rtps/xmlparser/XMLElementParser.cpp [hunk 8]
```diff
@@ -4409,8 +4410,75 @@ XMLP_ret XMLParser::getXMLThreadSettings(
...
+                    thread_setting.scheduling_policy < -1)
+            {
+                return XMLP_ret::XML_ERROR;
+            }
+        }
+        else if (strcmp(current_elem->Name(), PRIORITY) == 0)
+        {
+            // priority - int32Type
+            if (XMLP_ret::XML_OK != getXMLInt(current_elem, &thread_setting.priority, ident) ||
+                    thread_setting.priority < -1)
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply src/cpp/rtps/xmlparser/XMLElementParser.cpp]: 59e3e626a

File: test/unittest/xmlparser/XMLProfileParserTests.cpp [hunk 9]
```diff
@@ -3188,6 +3194,88 @@ TEST_F(XMLProfileParserBasicTests, external_locators_feature)
...
+                            <stack_size>-1</stack_size>
+                            <wrong_tag>0</wrong_tag>
+                        </thread_settings>
+                    </log>
+                </dds>)",
+            xmlparser::XMLP_ret::XML_ERROR
+        },
+    };
+
+    EXPECT_CALL(*log_mock, SetThreadConfig()).Times(2);
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply test/unittest/xmlparser/XMLProfileParserTests.cpp]: 59e3e626a

File: src/cpp/rtps/xmlparser/XMLParser.cpp [hunk 10]
```diff
@@ -1439,6 +1452,19 @@ XMLP_ret XMLParser::parseLogConfig(
             {
                 ret = parseXMLConsumer(*p_element);
             }
+            else if (strcmp(tag, THREAD_SETTINGS) == 0)
+            {
+                fastdds::rtps::ThreadSettings thread_settings;
+                ret = getXMLThreadSettings(*p_element, thread_settings);
+                if (ret == XMLP_ret::XML_OK)
+                {
+                    fastdds::dds::Log::SetThreadConfig(thread_settings);
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply src/cpp/rtps/xmlparser/XMLParser.cpp]: 59e3e626a

File: include/fastrtps/xmlparser/XMLProfileManager.h [hunk 11]
```diff
@@ -135,6 +138,22 @@ class XMLProfileManager
     RTPS_DllAPI static void getDefaultParticipantAttributes(
             ParticipantAttributes& participant_attributes);
 
+    /**
+     * Search for the profile specified and fill the structure.
+     * @param profile_name Name for the profile to be used to fill the structure.
+     * @param qos Structure to be filled.
+     * @param log_error Flag to log an error if the profile_name is not found.
+     * @return XMLP_ret::XML_OK on success, XMLP_ret::XML_ERROR in other case. Defaults true.
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply include/fastrtps/xmlparser/XMLProfileManager.h]: 59e3e626a

File: include/fastrtps/xmlparser/XMLProfileManager.h [hunk 12]
```diff
@@ -135,6 +136,22 @@ class XMLProfileManager
...
+     */
+    RTPS_DllAPI static XMLP_ret fillDomainParticipantFactoryQos(
+            const std::string& profile_name,
+            fastdds::dds::DomainParticipantFactoryQos& qos,
+            bool log_error = true);
+
+    //!Fills qos with the default values.
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply include/fastrtps/xmlparser/XMLProfileManager.h]: 59e3e626a

File: src/cpp/rtps/xmlparser/XMLParser.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (EduPonz) [reply src/cpp/rtps/xmlparser/XMLParser.cpp:2224]: 59e3e626a

File: test/unittest/xmlparser/XMLParserTests.cpp [hunk 13]
```diff
@@ -2876,6 +2880,133 @@ TEST_F(XMLParserTests, parseXMLTopicDataNegativeClauses)
...
+                    </wrong_tag>
+                </reception_threads>)",
+            xmlparser::XMLP_ret::XML_ERROR,
+            {}
+        },
+        {
+            "reception_threads_wrong_attribute",
+            R"(
+                <reception_threads>
+                    <reception_thread wrong_attribute="12345">
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply test/unittest/xmlparser/XMLParserTests.cpp:3068]: 59e3e626a

File: resources/xsd/fastRTPS_profiles.xsd [hunk 14]
```diff
@@ -1739,6 +1741,29 @@
...
+        | ╠ att. port        [uint32]
+        |
+        ├ scheduling_policy  [int32],
+        ├ priority           [int32],
+        ├ affinity           [uint64],
+        └ stack_size         [int32]-->
+    <xs:complexType name="threadSettingsWithPortType">
+        <xs:complexContent>
+            <xs:extension base="threadSettingsType">
+                <xs:attribute name="port" type="uint32" use="optional"/>
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply resources/xsd/fastRTPS_profiles.xsd]: 59e3e626a

File: src/cpp/fastdds/subscriber/DataReaderImpl.cpp [hunk 15]
```diff
@@ -1552,6 +1552,12 @@ bool DataReaderImpl::can_qos_be_updated(
         EPROSIMA_LOG_WARNING(RTPS_QOS_CHECK,
                 "Data sharing configuration cannot be changed after the creation of a DataReader.");
     }
+    if (!(to.data_sharing().data_sharing_listener_thread() == from.data_sharing().data_sharing_listener_thread()))
+    {
+        updatable = false;
+        EPROSIMA_LOG_WARNING(RTPS_QOS_CHECK,
+                "data_sharing_listener_thread cannot be changed after the DataReader is enabled");
+    }
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply src/cpp/fastdds/subscriber/DataReaderImpl.cpp]: 59e3e626a

File: resources/xsd/fastRTPS_profiles.xsd (hunk 1 above)
Comment (EduPonz) [reply resources/xsd/fastRTPS_profiles.xsd]: 59e3e626a

File: test/unittest/xmlparser/XMLElementParserTests.cpp [hunk 16]
```diff
@@ -4112,3 +4124,205 @@ TEST_F(XMLParserTests, env_var_substitution)
...
+{
+    /* Define the test cases */
+    std::vector<std::pair<std::vector<std::string>, XMLP_ret>> test_cases =
+    {
+        {{"12", "12", "12", "12", ""}, XMLP_ret::XML_OK},
+        {{"-1", "12", "12", "12", ""}, XMLP_ret::XML_OK},
+        {{"12", "-1", "12", "12", ""}, XMLP_ret::XML_OK},
+        {{"12", "12", "12", "-1", ""}, XMLP_ret::XML_OK},
+        {{"-2", "12", "12", "12", ""}, XMLP_ret::XML_ERROR},
+        {{"12", "12", "12", "-2", ""}, XMLP_ret::XML_ERROR},
```
Comment (MiguelCompany) [inline test/unittest/xmlparser/XMLElementParserTests.cpp:4141]: Checking for a negative affinity still pending:
```suggestion
        {{"-2", "12", "12", "12", ""}, XMLP_ret::XML_ERROR},
        {{"12", "12", "-2", "12", ""}, XMLP_ret::XML_ERROR},
        {{"12", "12", "12", "-2", ""}, XMLP_ret::XML_ERROR},
```

File: src/cpp/rtps/xmlparser/XMLParser.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/rtps/xmlparser/XMLParser.cpp:2224]: My suggestion was not followed. I suggest adding a warning when we ignore the tag.

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

File: test/unittest/xmlparser/XMLElementParserTests.cpp (hunk 16 above)
Comment (EduPonz) [reply test/unittest/xmlparser/XMLElementParserTests.cpp]: a807237

File: src/cpp/rtps/xmlparser/XMLParser.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (EduPonz) [reply src/cpp/rtps/xmlparser/XMLParser.cpp]: a807237

**Review by MiguelCompany — APPROVED** [review]
LGTM

**Comment by MiguelCompany** [comment]
@richiprosima Please test this

**Comment by EduPonz** [comment]
@richiprosima please test mac

**Comment by MiguelCompany** [comment]
@richiprosima Please test this

**Comment by MiguelCompany** [comment]
@richiprosima Please test this

**Comment by MiguelCompany** [comment]
@richiprosima Please test mac

**Comment by MiguelCompany** [comment]
@richiprosima Please test mac

**Comment by EduPonz** [comment]
@richiprosima please test mac

**Review by MiguelCompany — APPROVED** [review]
